File: facade/java/FacadePatternDemo.java
interface Shape {

	void draw();
}

final class Circle implements Shape {

	@Override
	public void draw() {
		System.out.println("Circle::draw");
	}
}

final class Rectangle implements Shape {

	@Override
	public void draw() {
		System.out.println("Rectangle::draw");
	}
}

final class Square implements Shape {

	@Override
	public void draw() {
		System.out.println("Square::draw");
	}
}

final class ShapeFacade {

	private final Shape circle, square, rectangle;

	public ShapeFacade() {
		this.circle = new Circle();
		this.square = new Square();
		this.rectangle = new Rectangle();
	}

	public void drawCircle() {
		circle.draw();
	}

	public void drawRectangle() {
		rectangle.draw();
	}

	public void drawSquare() {
		square.draw();
	}
}

final class FacadePatternDemo {

	public static void main(String[] args) {
		ShapeFacade shapeFacade = new ShapeFacade();
		shapeFacade.drawSquare();
		shapeFacade.drawCircle();
		shapeFacade.drawRectangle();
	}
}

File: facade/kotlin/facade_pattern_demo.kt
interface Shape {

	fun draw(): Unit
}

class Circle : Shape {

	override fun draw(): Unit {
		println("Circle::draw")
	}
}

class Rectangle : Shape {

	override fun draw(): Unit {
		println("Rectangle::draw")
	}
}

class Square : Shape {

	override fun draw(): Unit {
		println("Square::draw")
	}
}

class FacadePattern(
	private val circle: Shape,
	private val square: Shape,
	private val rectangle: Shape,
) {

	constructor() : this(Circle(), Square(), Rectangle())

	fun drawCircle(): Unit {
		this.circle.draw();
	}

	fun drawRectangle(): Unit {
		this.rectangle.draw();
	}

	fun drawSquare(): Unit {
		this.square.draw();
	}
}

fun main(): Unit {
	val fp: FacadePattern = FacadePattern()
	fp.drawCircle();
	fp.drawRectangle();
	fp.drawSquare();
}